proxy/src/coco.rs

```diff
@@ -1,649 +1,457 @@
 //! Abstractions and utilities for git interactions through the API.
 
-use std::collections::hash_map::DefaultHasher;
+use async_trait::async_trait;
 use std::env;
-use std::fmt;
-use std::hash::{Hash, Hasher};
 use std::str::FromStr;
 
 use librad::git;
+use librad::git::storage::Storage;
 use librad::keys;
-use librad::meta;
+use librad::meta::entity;
+use librad::meta::project;
+use librad::meta::user;
 use librad::paths::Paths;
-use librad::peer;
-use librad::project;
 use librad::surf;
 use librad::surf::git::git2;
+use librad::uri::{self, RadUrn};
+use radicle_keystore::{Keypair, Keystore, SecretKeyExt};
 
 use crate::error;
 
-/// Branch name representation.
-#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
-pub struct Branch(String);
-
-impl fmt::Display for Branch {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}", self.0)
-    }
+mod source;
+pub use source::{
+    blob, branches, commit, commits, local_branches, tags, tree, Blob, BlobContent, Branch, Commit,
+    Info, ObjectType, Person, Tag, Tree, TreeEntry,
+};
+
+/// The set of capabilities necessary for interacting with `radicle-link`.
+#[async_trait]
+pub trait Client:
+    Keystore<
+        PublicKey = keys::PublicKey,
+        SecretKey = keys::SecretKey,
+        Metadata = <keys::SecretKey as SecretKeyExt>::Metadata,
+        Error = error::Error,
+    > + entity::Resolver<project::Project>
+    + entity::Resolver<user::User>
+    + Send
+    + Sync
+{
+    /// Get the [`git::repo::Repo`] for the given `project_urn`.
+    async fn get_repo(
+        &self,
+        project_urn: String,
+    ) -> Result<(git::repo::Repo, project::Project), error::Error>;
+
+    /// Initialize a [`librad::project::Project`] in the location of the given `path`.
+    ///
+    /// # Errors
+    ///
+    /// Will return [`error::Error`] if the git2 repository is not present for the `path` or any of
+    /// the librad interactions fail.
+    fn init_project(
+        &self,
+        owner: &user::User,
+        path: &str,
+        name: &str,
+        description: &str,
+        default_branch: &str,
+    ) -> Result<(RadUrn, project::Project), error::Error>;
+
+    /// Returns the list of [`librad::project::Project`] known for the configured [`Paths`].
+    #[must_use]
+    fn list_projects(&self) -> Vec<(RadUrn, project::Project)>;
+
+    /// Create a copy of the git-platinum repo, init with coco and push tags and the additional dev
+    /// branch.
+    ///
+    /// # Errors
+    ///
+    /// Will return [`error::Error`] if any of the git interaction fail, or the initialisation of
+    /// the coco project.
+    fn replicate_platinum(
```

Comment:
This seems like it shouldn't be in the client. It can be defined in terms of `init_project`, I think.